rasa/core/featurizers/precomputation.py

```diff
@@ -0,0 +1,462 @@
+                # attribute even if the corresponding feature list is empty.
+                if feat_value and existing_values:
+                    raise RuntimeError(
+                        f"Feature for attribute {feat_attribute} has already been "
+                        f"extracted from a different message stored under a key "
+                        f"in {key_attributes} "
+                        f"that is different from {key_attribute}. This means there's a "
+                        f"redundancy in the lookup table."
+                    )
+                if feat_value:
+                    features[feat_attribute] = feat_value
+        return features
+
+    def lookup_message(self, user_text: Text) -> Message:
+        """Returns the lookup table entry identified by the given user text.
+
+        Args:
+          user_text: the text of a user utterance
+        Raises:
+          `ValueError` if there is no message associated with the given user text
+        """
+        key = self._build_key({TEXT: user_text})
+        message = self._table.get(key)
+        if message is None:
+            raise ValueError(f"Expected a message with key {key} in lookup table.")
+        return message
+
+    def derive_messages_from_domain_and_add(self, domain: Domain) -> None:
+        """Adds all lookup table entries that can be derived from the domain.
+
+        Args:
+          domain: the domain from which we extract the substates
+          lookup_table: lookup table to which the substates will be added (as messages)
+        """
+        action_texts = domain.action_texts
+        action_names = domain.action_names_or_texts[
+            slice(0, -len(domain.action_texts) if domain.action_texts else None)
+        ]
+        if set(action_texts).intersection(action_names):
+            raise NotImplementedError(
+                "We assumed that domain's action_names_or_texts contains all "
+                "action names followed by the texts. Apparently that changed. "
+            )
+        for tag, actions in [(ACTION_NAME, action_names), (ACTION_TEXT, action_texts)]:
+            for action in actions:
+                self.add(Message({tag: action}))
+
+        for intent in domain.intent_properties.keys():
+            self.add(Message({INTENT: intent}))
+
+    def derive_messages_from_events_and_add(self, events: Iterable[Event]) -> None:
+        """Creates all possible action and (partial) user substates from the events.
+
+        Note that partial user substate means user substate with intent only, user
+        substate with all attributes expect intent, "prev_action" substate with
+        action text only or "prev_action" substate with action name only.
+
+        Args:
+          events: list of events to extract the substate from
+          lookup_table: lookup table to which the substates will be added (as messages)
+        """
+        for event in events:
+            if isinstance(event, UserUttered):
```

Comment:
You mean the partial state? There is a bit (i.e. too much) text in the class annotation that is supposed to explain this:
```
The motivation for this lookup table is that
    a) our policies only need to pass very specific information through the
    tokenization and featurization pipeline
    b) our tokenization and featurization pipeline featurize certain attributes
    independently of each other.
 ```
 I could remove hte "partial substate ... blabla" and just add a "see class anotation" kind of comment?

Reply:
Kind of 😆 I thought of something like `Policies with end-to-end support featurize conversations twice: one time assuming that the intent should be used for prediction and one time assuming the message text should be used for featurization. Hence we need to featurize two states for 'UserUttered' events`. Does that make sense?

Reply:
I think the indepdence of the featurizations is the stronger point here because the "featurize once with and once without text" would be confusing otherwise and because I could easily imagine algorithms using both. - But I can rearrange that text and add that explanation on what happens with those features and that event could result in two different (sub-states) :)

Reply:
Thank you! 🙌🏻